### ISSUE TITLE
mxnet.ndarray.NDArray objects cannot be concatenated

### ISSUE THREAD

**Issue by rldotai** [dmlc/mxnet#1985]
I had some problems after converting my network from `mxnet.model.FeedForward` to `mxnet.module.Module` when using the `predict()` method.

With a module that has been trained, and a DataIterator `test_iter`, the following returns an error:

```python
mod.predict(test_iter)
```

The exception raised is:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-134-73b4dc77d8cf> in <module>()
----> 1 model.predict(test_iter)

python/mxnet/module/base_module.py in predict(self, eval_data, num_batch, merge_batches, reset, always_output_list)
    259                        'in mini-batches. Maybe bucketing is used?'
    260             output_list2 = [np.concatenate([out[i] for out in output_list])
--> 261                             for i in range(num_outputs)]

ValueError: zero-dimensional arrays cannot be concatenated
```

[The code in question appears here](https://github.com/dmlc/mxnet/blob/master/python/mxnet/module/base_module.py#L260)

The problem seems to be due to trying to concatenate `mxnet.ndarray.NDArray` objects, which appear to have dimension 0.

```python
>>> pred = model.predict(test_iter, merge_batches=False) # This works 
>>> np.ndim(pred[0][0]) 
0
# Checking that it's not due to end-of-batch-effects
>>> len(pred)
79
>>> pred[0]
[<mxnet.ndarray.NDArray at 0x1269f9240>]
>>> np.concatenate([pred[0][0], pred[1][0]]) 
...
ValueError: zero-dimensional arrays cannot be concatenated
```

There is no problem when using `mxnet.model.FeedForward.predict()`, so I took a look to see what was different between the two `predict()` functions. 

In the `FeedForward` one, outputs are converted to numpy arrays before concatenation, which solves the problem pretty easily.

[`mxnet.model.FeedForward.predict`](https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L619)

I will attach a minor change in lieu of rewriting the NDArray module to support concatenation.

**Comment by rldotai** [comment]
OK, I am unsure of how to get this to work since I already have a pull request pending, but there's some sort of issue with the linter so it hasn't been merged.
So now they appear to be two commits on the same pull request, #1977

The change I suggested was for `mxnet/python/mxnet/module/base_module.py`, altering line 247:

```
-            outputs = [out[0:out.shape[0]-pad] for out in self.get_outputs()]
+            outputs = [out[0:out.shape[0]-pad].asnumpy() for out in self.get_outputs()]
```

I can attest that it fixed my issues using `predict` and seemed like the easiest change that would ensure similar behavior.

**Comment by pluskid** [comment]
@rldotai Thanks for spotting this bug! I proposed a different fix in #1989, which still keeps the `NDArray` as output types, but call `nd.concatenate` instead of `np.concatenate`. Returning `NDArray` allows us to better utilize the asynchronized execution engine in the backend and also a relatively consistent API. Please call `asnumpy()` explicitly on the returned results if you need to work with numpy. Thanks again!

**Comment by rldotai** [comment]
No worries, just glad to help.
